crates/uv-publish/src/lib.rs

```diff
@@ -0,0 +1,784 @@
+    /// Tokens or Trusted Publishers instead. See https://test.pypi.org/help/#apitoken and
+    /// https://test.pypi.org/help/#trusted-publishers",
+    /// "title": "Forbidden"}
+    /// ```
+    ///
+    /// For context, for the last case twine shows:
+    /// ```text
+    /// WARNING  Error during upload. Retry with the --verbose option for more details.
+    /// ERROR    HTTPError: 403 Forbidden from https://test.pypi.org/legacy/
+    ///          Username/Password authentication is no longer supported. Migrate to API
+    ///          Tokens or Trusted Publishers instead. See
+    ///          https://test.pypi.org/help/#apitoken and
+    ///          https://test.pypi.org/help/#trusted-publishers
+    /// ```
+    ///
+    /// ```text
+    /// INFO     Response from https://test.pypi.org/legacy/:
+    ///          403 Username/Password authentication is no longer supported. Migrate to
+    ///          API Tokens or Trusted Publishers instead. See
+    ///          https://test.pypi.org/help/#apitoken and
+    ///          https://test.pypi.org/help/#trusted-publishers
+    /// INFO     <html>
+    ///           <head>
+    ///            <title>403 Username/Password authentication is no longer supported.
+    ///          Migrate to API Tokens or Trusted Publishers instead. See
+    ///          https://test.pypi.org/help/#apitoken and
+    ///          https://test.pypi.org/help/#trusted-publishers</title>
+    ///           </head>
+    ///          <body>
+    ///           <h1>403 Username/Password authentication is no longer supported.
+    ///         Migrate to API Tokens or Trusted Publishers instead. See
+    ///          https://test.pypi.org/help/#apitoken and
+    ///          https://test.pypi.org/help/#trusted-publishers</h1>
+    ///            Access was denied to this resource.<br/><br/>
+    /// ```
+    ///
+    /// In comparison, we now show (line-wrapped for readability):
+    ///
+    /// ```text
+    /// error: Failed to publish `dist/astral_test_1-0.1.0-py3-none-any.whl` to `https://test.pypi.org/legacy/`
+    ///   Caused by: Incorrect credentials (status code 403 Forbidden): 403 Username/Password
+    ///     authentication is no longer supported. Migrate to API Tokens or Trusted Publishers
+    ///     instead. See https://test.pypi.org/help/#apitoken and https://test.pypi.org/help/#trusted-publishers
+    /// ```
+    fn extract_error_message(body: String, content_type: Option<&str>) -> String {
+        if content_type == Some("application/json") {
+            #[derive(Deserialize)]
+            struct ErrorBody {
+                code: String,
+            }
+
+            if let Ok(structured) = serde_json::from_str::<ErrorBody>(&body) {
+                structured.code
+            } else {
+                body
+            }
+        } else {
+            body
+        }
+    }
+}
+
+pub fn files_for_publishing(
+    paths: Vec<String>,
```

Comment:
Nit: I don't think this needs to be owned, it could even be an iterator over `&str`.